returnn/tf/layers/basic.py

```diff
@@ -2695,22 +2693,41 @@ def get_out_data_from_opts(cls, name, sources, axis="T", **kwargs):
     # Just same format.
     return get_concat_sources_data_template(sources, name="%s_output" % name)
 
+  # noinspection PyMethodOverriding
   @classmethod
-  def get_rec_initial_extra_outputs(cls, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
+  def get_rec_initial_extra_outputs(cls, network, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
     """
+    :param returnn.tf.network.TFNetwork network:
     :param tf.Tensor batch_dim:
     :param returnn.tf.layers.rec.RecLayer|LayerBase rec_layer:
     :param str axis:
     :param list[LayerBase] sources:
     :rtype: dict[str,tf.Tensor]
     """
     data = get_concat_sources_data_template(sources)
-    if axis == "T" and data.time_dim_axis is None:
-      # Assume inside RecLayer.
+    if cls._is_recurrent(axis=axis, network=network, data=data):
       assert all(data.shape)
       return {"state": tf.zeros(data.get_batch_shape(batch_dim=batch_dim), dtype=data.dtype)}
     return {}
 
+  @classmethod
+  def _is_recurrent(cls, axis, network, data):
+    """
+    :param str|DimensionTag axis:
+    :param returnn.tf.network.TFNetwork network:
+    :param Data data:
+    :rtype: bool
+    """
+    if not network.is_inside_rec_layer(inside_loop=False):
+      return False
+    if axis.lower() == "t" and data.time_dim_axis is None:  # legacy fallback
+      return True
+    if axis.lower() == ":i":
+      return True
+    if not isinstance(axis, DimensionTag):
+      return False
+    return axis == network.get_rec_parent_layer(inside_loop=False).time_dim_tag
```

Comment:
This misses some check if a rec parent layer exists. Otherwise this returns `None` and you get an attribute error here.

But some code like this would be better anyway:
```suggestion
    return axis == network.get_inside_rec_time_dim()
```